### ISSUE TITLE
cgroup backend improvements

### ISSUE THREAD

**Issue by brauner** [opencontainers/runc#892]
Currently runC finds the cgroup root mountpoint e.g. /sys/fs/cgroup for the
cgroupfs. It then finds the first prefix following the rount mountpoint for a
given subsystem. Assume our user's shell is residing in the following cgroups in
a mounted hierarchy:

```
cat /proc/self/cgroup

11:hugetlb:/
10:blkio:/user.slice
9:net_cls,net_prio:/user.slice
8:cpuset:/
7:freezer:/
6:devices:/user.slice
5:pids:/user.slice/user-1000.slice/session-1.scope
4:memory:/user.slice
3:cpu,cpuacct:/user.slice
2:perf_event:/
1:name=systemd:/user.slice/user-1000.slice/session-1.scope
```

Currently runc will take the first prefix following a subsystem name and
place the container into it for *all* subsystems. For the above example this
means that runC will find e.g. the prefix `/user.slice`. If we now have a
container called `test` then the container will be placed into
`/sys/fs/cgroup/*/user.slice/test` for *all* subsystems:

```
cat /proc/self/cgroup

11:hugetlb:/user.slice/test
10:blkio:/user.slice/test
9:net_cls,net_prio:/user.slice/test
8:cpuset:/user.slice/test
7:freezer:/user.slice/test
6:devices:/user.slice/test
5:pids:/user.slice/test
4:memory:/user.slice/test
3:cpu,cpuacct:/user.slice/test
2:perf_event:/user.slice/test
1:name=systemd:/user.slice/test
```

This seems to be a bug. In the case of a privileged container (privC) this is
not a problem since we have write access to all mounted subsystems anyway but it
is still strange.  The expected behaviour should rather be similar to other
cgroup backends (e.g.  as seen in Docker or LXC) where our privC "test" is
placed in a cgroup in the root mount for each subsystem.

```
  Docker:

11:hugetlb:/docker/test
10:blkio:/docker/test
9:net_cls,net_prio:/docker/test
8:cpuset:/docker/test
7:freezer:/docker/test
6:devices:/docker/test
5:pids:/docker/test
4:memory:/docker/test
3:cpu,cpuacct:/docker/test
2:perf_event:/docker/test
1:name=systemd:/docker/test
```

```
LXC

11:hugetlb:/lxc/test
10:blkio:/lxc/test/user.slice/user-0.slice/session-1.scope
9:net_cls,net_prio:/lxc/test
8:cpuset:/lxc/test
7:freezer:/lxc/test
6:devices:/lxc/test/user.slice/user-0.slice/session-1.scope
5:pids:/lxc/test
4:memory:/lxc/test/user.slice/user-0.slice/session-1.scope
3:cpu,cpuacct:/lxc/test/user.slice/user-0.slice/session-1.scope
2:perf_event:/lxc/test
1:name=systemd:/lxc/test/user.slice/user-0.slice/session-1.scope

```

In the case of an unprivileged container (unprivC) the current runC cgroup backend
 becomes a bug since we might be placed in a writeable cgroup for some subsystems
while being placed in a non-writeable cgroup for other subsystems. What runC should
probably do is to find the mountpoint for each subsystem individually and record the
prefix. I've talked to @cyphar and I'll start working on rewriting the cgroup backend if you're ok with that.

**Comment by cyphar** [comment]
It should be noted that, in my opinion, this is a legitimate bug in our automated generation of `--cgroup-path` (a user wouldn't expect the default of setting up a container to be putting it in essentially a random cgroup -- as opposed to putting it in a child cgroup of the current one).

I'm adding this to the v1.0.0 milestone.

/cc @crosbymichael @mrunalp

**Comment by crosbymichael** [comment]
This was resolved by #1009
